psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -888,24 +980,436 @@ void DirectJK::build_JK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, std::v
+        size_t ntri = nshell * (nshell + 1L) / 2L;
+        size_t possible_shells = ntri * (ntri + 1L) / 2L;
+        printer.Printf("Computed %20zu Shell Quartets out of %20zu, (%11.3E ratio)\n", computed_shells,
+                        possible_shells, computed_shells / (double)possible_shells);
+    }
+
+    timer_off("build_JK_matrices()");
+}
+
+// To follow this code, compare with figure 1 of DOI: 10.1063/1.476741
+void DirectJK::build_linK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, const std::vector<SharedMatrix>& D,
+                  std::vector<SharedMatrix>& K) {
+
+    if (!lr_symmetric_) {
+        throw PSIEXCEPTION("Non-symmetric K matrix builds are currently not supported in the LinK algorithm.");
+    }
+
+    timer_on("build_linK()");
+
+    // ==> Prep Auxiliary Quantities <== //
+
+    // => Zeroing <= //
+    for (auto& Kmat : K) {
+        Kmat->zero();
+    }
+
+    // => Sizing <= //
+    int nshell = primary_->nshell();
+    int nbf = primary_->nbf();
+    int nthread = df_ints_num_threads_;
+
+    // => Atom Blocking <= //
+    std::vector<int> shell_endpoints_for_atom;
+    std::vector<int> basis_endpoints_for_shell;
+
+    int atomic_ind = -1;
+    for (int P = 0; P < nshell; P++) {
+        if (primary_->shell(P).ncenter() > atomic_ind) {
+            shell_endpoints_for_atom.push_back(P);
+            atomic_ind++;
+        }
+        basis_endpoints_for_shell.push_back(primary_->shell_to_basis_function(P));
+    }
+    shell_endpoints_for_atom.push_back(nshell);
+    basis_endpoints_for_shell.push_back(nbf);
+
+    size_t natom = shell_endpoints_for_atom.size() - 1;
+
+    size_t max_functions_per_atom = 0L;
+    for (size_t atom = 0; atom < natom; atom++) {
+        size_t size = 0L;
+        for (int P = shell_endpoints_for_atom[atom]; P < shell_endpoints_for_atom[atom + 1]; P++) {
+            size += primary_->shell(P).nfunction();
+        }
+        max_functions_per_atom = std::max(max_functions_per_atom, size);
+    }
+
+    if (debug_) {
+        outfile->Printf("  ==> LinK: Atom Blocking <==\n\n");
+        for (size_t atom = 0; atom < natom; atom++) {
+            outfile->Printf("  Atom: %3d, Atom Start: %4d, Atom End: %4d\n", atom, shell_endpoints_for_atom[atom],
+                            shell_endpoints_for_atom[atom + 1]);
+            for (int P = shell_endpoints_for_atom[atom]; P < shell_endpoints_for_atom[atom + 1]; P++) {
+                int size = primary_->shell(P).nfunction();
+                int off = primary_->shell(P).function_index();
+                int off2 = basis_endpoints_for_shell[P];
+                outfile->Printf("    Shell: %4d, Size: %4d, Offset: %4d, Offset2: %4d\n", P, size, off,
+                                off2);
+            }
+        }
+        outfile->Printf("\n");
+    }
+
+    // ==> Prep Atom Pairs <== //
+    // Based on limited evidence, atom blocking is better for parallel performance than shell-pair blocking
```

Comment:
This comment is still inaccurate, I am unaware of any evidence that shows this. Please change to something like: 
```
// atom-pair blocking inherited from DirectJK code
// TODO: test shell-pair blocking
```

Reply:
I _still_ disagree with your proposed changes. I suspect that Andy worded it this way to try and compromise between my recommendation and your recommendation, in the interests of getting this PR over and done with. The best course of action seems to be to settle this between the two of us.

The way I understand it, the central question is _to what extent was shell-pair screening tested_? Andy has been vague on this point, and I'm not convinced that the answer to that question is "none." Andy discussed the matter with CDS [much earlier in the review process](https://github.com/psi4/psi4/pull/2359#issuecomment-998996631), and the two of them decided atom-pair blocking was likely to be better. I know that Andy ran tests which found shell-blocking led to ["severely impaired" loop efficiency](https://github.com/psi4/psi4/pull/2359#issuecomment-1003741024), although it is correct that we don't know how much effort went into code optimization.

@andyj10224 **Exactly** why did you and CDS think atom-blocking was likely to be more efficient than shell-blocking, and **exactly** what kinds of tests did you run on shell-blocking parallelization and how severe is "severe"? Giving an honest and detailed answer is the best way to end this argument.

Reply:
@JonathonMisiewicz I implemented a variant of LinK that did shell-blocking, and I found that for a few small test systems (water clusters), it was systematically slower than the atom-blocking variant of LinK. Of course, in hindsight, I could have easily changed the parallel scheduling algorithm to better optimize my shell-pair algorithm, so it was comparing atom-blocking to an `unoptimized` shell_pair algorithm. Since those were the extent of the tests, I actually agree with Zach's changes.

Reply:
That answers my second question, but what about the other two?

How severe was the effect, and what about the conversation with CDS where he said that shell-blocking was a bad idea?

Reply:
It was not very severe, but it was noticeable, a 7-10 second difference was noticed (out of a wall time of around 2 minutes). CDS's worry was that we would have "too-small workloads in the inner loops", hindering parallel performance.

Reply:
I'm not an expert in parallelization, but I'm imagining that's within the threshold of "optimization could fix this."

If shell-pair blocking is a reasonable thing to try, then go ahead and add the comment per Zach's suggestion. This is increasingly sounding like the case for atom-pair blocking was oversold earlier in the review process...

Reply:
Agreed, the biggest reason why I used atom-blocking was that it was borrowed from the DirectJK loops.

Shell-blocking is actually worth trying, I use shell-blocking in my current CFMM code and it is actually very efficient. Atom-blocking vs shell-blocking is an argument outside the scope of this PR and may require a major refactor of the DirectJK loop itself.